packages/@n8n/task-runner/src/task-runner.ts

```diff
@@ -284,111 +286,116 @@ export abstract class TaskRunner extends EventEmitter {
 		}
 
 		this.resetIdleTimer();
-		this.runningTasks.set(taskId, {
+		const task = new TaskState({
 			taskId,
-			active: false,
-			cancelled: false,
+			timeoutInS: this.taskTimeout,
+			onTimeout: () => {
+				void this.taskTimedOut(taskId);
+			},
 		});
+		this.runningTasks.set(taskId, task);
 
 		this.send({
 			type: 'runner:taskaccepted',
 			taskId,
 		});
 	}
 
-	taskCancelled(taskId: string, reason: string) {
-		const task = this.runningTasks.get(taskId);
-		if (!task) {
+	async taskCancelled(taskId: string, reason: string) {
+		const taskState = this.runningTasks.get(taskId);
+		if (!taskState) {
 			return;
 		}
-		task.cancelled = true;
 
-		for (const [requestId, request] of this.dataRequests.entries()) {
-			if (request.taskId === taskId) {
-				request.reject(new TaskCancelledError(reason));
-				this.dataRequests.delete(requestId);
-			}
-		}
+		await taskState.caseOf({
+			// If the cancelled task hasn't received settings yet, we can finish it
+			waitingForSettings: () => this.finishTask(taskState),
 
-		for (const [requestId, request] of this.nodeTypesRequests.entries()) {
-			if (request.taskId === taskId) {
-				request.reject(new TaskCancelledError(reason));
-				this.nodeTypesRequests.delete(requestId);
-			}
-		}
+			// If the task has already timed out or is already cancelled, we can
+			// ignore the cancellation
+			'aborting:timeout': noOp,
+			'aborting:cancelled': noOp,
 
-		const controller = this.taskCancellations.get(taskId);
-		if (controller) {
-			controller.abort();
-			this.taskCancellations.delete(taskId);
+			running: () => {
+				taskState.status = 'aborting:cancelled';
+				taskState.abortController.abort('cancelled');
+				this.cancelTaskRequests(taskId, reason);
+			},
+		});
+	}
+
+	async taskTimedOut(taskId: string) {
+		const taskState = this.runningTasks.get(taskId);
+		if (!taskState) {
+			return;
 		}
 
-		if (!task.active) this.runningTasks.delete(taskId);
+		await taskState.caseOf({
+			// If we are still waiting for settings for the task, we can error the
+			// task immediately
+			waitingForSettings: () => {
+				this.send({
+					type: 'runner:taskerror',
+					taskId,
+					error: new TimeoutError(this.taskTimeout),
+				});
 
-		this.sendOffers();
-	}
+				this.finishTask(taskState);
+			},
```

Comment:
If `this.send` throws because e.g. the websocket connection closed, we won't clean up in `finishTask`?

Reply:
Good catch! I'll wrap it in try..finally